### PR TITLE
small perf improvements

### PR DESCRIPTION
**Pull request by sfc-gh-chu**
# Description

 - Fix app caching and updates sqlalchemy bulk insert api.
 - Call stack related speedups

## Other details good to know for developers

Please include any other details of this change useful for _TruLens_ developers.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to
  not work as expected)
- [ ] New Tests
- [ ] This change includes re-generated golden test results
- [ ] This change requires a documentation update

### CODE REVIEW

File: src/core/trulens/core/schema/app.py [hunk 1]
```diff
@@ -352,6 +352,8 @@ def _submit_feedback_functions(
         tp = TP()
 
         for ffunc in feedback_functions:
+            ffunc.run(app=app, record=record)
```
Comment (sfc-gh-pdharmana) [inline src/core/trulens/core/schema/app.py:355]: Why is this needed?

File: src/core/trulens/core/database/sqlalchemy.py [hunk 2]
```diff
@@ -378,7 +378,7 @@ def batch_insert_record(
                 self.orm.Record.parse(r, redact_keys=self.redact_keys)
                 for r in records
             ]
-            session.bulk_save_objects(records_list)
+            session.add_all(records_list)
```
Comment (sfc-gh-pdharmana) [inline src/core/trulens/core/database/sqlalchemy.py:381]: Whats difference between add_all and bulk_save objects?

File: src/core/trulens/core/schema/app.py (hunk 1 above)
Comment (sfc-gh-chu) [reply src/core/trulens/core/schema/app.py:355]: Ah some working changes got checked in. Will revert this

File: src/core/trulens/core/database/sqlalchemy.py (hunk 2 above)
Comment (sfc-gh-chu) [reply src/core/trulens/core/database/sqlalchemy.py:381]: Functionally the same thing. [bulk_save_objects](https://docs.sqlalchemy.org/en/20/orm/session_api.html#sqlalchemy.orm.Session.bulk_save_objects) is the legacy API in SQLAlchemy 2.0 so updated this to use the `add_all` API.

File: src/core/trulens/core/utils/python.py [hunk 3]
```diff
@@ -535,37 +543,43 @@ def merge_stacks(
...
             pass
 
     return ret
 
 
-def stack_with_tasks() -> Sequence[FrameType]:
+def stack_generator(offset: int = 0) -> Iterable[FrameType]:
+    frame = inspect.currentframe()
+    for _ in range(offset):
+        assert frame is not None
```
Comment (sfc-gh-dhuang) [inline src/core/trulens/core/utils/python.py:562]: Do we have to use assert here? or can we just raise exception with error message

File: src/core/trulens/core/utils/python.py [hunk 4]
```diff
@@ -437,10 +445,10 @@ def external_caller_frame(offset=0) -> FrameType:
     Raises:
         RuntimeError: If no such frame is found.
     """
-
-    for finfo in inspect.stack()[offset + 1 :]:
-        if not finfo.frame.f_globals["__name__"].startswith("trulens"):
-            return finfo.frame
+    gen = list(stack_generator(offset + 2))
```
Comment (sfc-gh-dhuang) [inline src/core/trulens/core/utils/python.py:448]: casting to list shouldn't be necessary here right? and it's more memory efficient if you just iterate over the generator without forcing to evaluate the entire stack all at once

**Comment by sfc-gh-dhuang** [comment]
left a few comments but minor. mostly LGTM

File: src/core/trulens/core/utils/python.py (hunk 4 above)
Comment (sfc-gh-chu) [reply src/core/trulens/core/utils/python.py:448]: I was originally thinking the same. Then realized that because the generator is called lazily, its call stack would change and reflect the generator callee's call stack, not the call stack of `external_caller_frame`

Comment (sfc-gh-dhuang) [reply src/core/trulens/core/utils/python.py:448]: ah got it sg

**Review by sfc-gh-dhuang — APPROVED** [review]

File: src/core/trulens/core/utils/python.py (hunk 4 above)
Comment (sfc-gh-chu) [reply src/core/trulens/core/utils/python.py:448]: actually we may be able to get around this by passing in the current frame so that inspect.currentframe() is guarenteed to be invoked at the right time
